File: problems/linked-list/0155-min-stack.js
/**
 * #155 Min Stack
 * Keep the minimum at each depth so getMin is constant time.
 * Each pushed value has a corresponding minimum-so-far entry.
 * Time: O(1) per operation | Space: O(n)
 */
class MinStack {
  constructor() {
    this.values = []
    this.minimums = []
  }

  push(value) {
    this.values.push(value)
    const currentMinimum = this.minimums[this.minimums.length - 1]
    this.minimums.push(
      currentMinimum === undefined ? value : Math.min(value, currentMinimum),
    )
  }

  pop() {
    this.minimums.pop()
    return this.values.pop()
  }

  top() {
    return this.values[this.values.length - 1]
  }

  getMin() {
    return this.minimums[this.minimums.length - 1]
  }
}

module.exports = MinStack

File: problems/linked-list/0155-min-stack.py
"""#155 Min Stack
Store the minimum at every stack depth for constant-time getMin.
All operations: O(1) | Space: O(n)
"""


class MinStack:
    def __init__(self):
        self.values = []
        self.minimums = []

    def push(self, val):
        self.values.append(val)
        current_minimum = self.minimums[-1] if self.minimums else val
        self.minimums.append(min(val, current_minimum))

    def pop(self):
        self.minimums.pop()
        self.values.pop()

    def top(self):
        return self.values[-1]

    def getMin(self):
        return self.minimums[-1]
